src/drivers/kinetis/uart.c

```diff
@@ -0,0 +1,269 @@
+ * @return number of bytes actually read. Number of bytes available
+ * if bytes == 0.
+ */
+unsigned int uart_read(uart_context *context, char *buf, unsigned int bytes);
+
+/**
+ * Not implemented. UARTs are not de-configured.
+ */
+unsigned int uart_close(uart_context *context);
+/* End UNIX state and prototypes */
+
+/**
+ * Interrupt Service Routines
+ *     state and prototypes
+ */
+/**
+ * Generic ISR for UART RX/TX interrupts
+ */
+void uart_isr(int which_uart);
+
+queue_t txq_headers[6];
+queue_t rxq_headers[6];
+volatile char txqs[6][255];
+volatile char rxqs[6][255];
+uart_context contexts[6] = {
+    {UART0, &txq_headers[0], &rxq_headers[0]},
+    {UART1, &txq_headers[1], &rxq_headers[1]},
+    {UART2, &txq_headers[2], &rxq_headers[2]},
+    {UART3, &txq_headers[3], &rxq_headers[3]},
+    {UART4, &txq_headers[4], &rxq_headers[4]},
+    {UART5, &txq_headers[5], &rxq_headers[5]}
+};
+/* End ISR state and prototypes */
+
+
+int uart_init(uart_config conf) {
+    //define variables for baud rate and baud rate fine adjust
+    uint16_t ubd, brfa;
+
+    // Enable clocks
+    *(conf.uart_clock_gate_base) |= conf.uart_clock_gate_mask;
+    *(conf.port_clock_gate_base) |= conf.port_clock_gate_mask;
+
+     
+    // Set port muxes
+    *(conf.tx_pcr) = conf.tx_alt << PORT_PCR_MUX_SHIFT;
+    *(conf.rx_pcr) = conf.rx_alt << PORT_PCR_MUX_SHIFT;
+    if(conf.rt_alt < 255) {
+        *(conf.rt_pcr) = conf.rt_alt << PORT_PCR_MUX_SHIFT;
+    }
+
+    if(conf.ct_alt < 255) {
+        *(conf.ct_pcr) = conf.ct_alt << PORT_PCR_MUX_SHIFT;
+    }
+
+    // Disable RX/TX until after configuration.
+    conf.uart_base->C2 &= ~UART_C2_RE_MASK;
+    conf.uart_base->C2 &= ~UART_C2_TE_MASK;
+
+    conf.uart_base->C1 = 0;
+
+
+    // baud rate = UART module clock / (16 � (SBR[12:0] + BRFD))
+    // BRFD is dependent on BRFA, refer to Table 52-234 in K64 reference manual
+    // BRFA is defined by the lower 4 bits of control register, UART0_C4 
+
+    //calculate baud rate settings: ubd = UART module clock/16* baud rate
+    ubd = (uint16_t)((conf.input_clock_rate)/(conf.baud * 16));
+
+    conf.uart_base->BDH &= ~UART_BDH_SBR_MASK;
+      
+    //distribute this ubd in BDH and BDL
+    conf.uart_base->BDH |= (UART_BDH_SBR_MASK & (ubd >> 8));
+    conf.uart_base->BDL = ubd;
+
+
+    //BRFD = (1/32)*BRFA 
+    //make the baud rate closer to the desired value by using BRFA
+    brfa = (((conf.input_clock_rate*32)/(conf.baud * 16)) - (ubd * 32));
+
+    conf.uart_base->C4 |= (UART_C4_BRFA_MASK & brfa);
+        
+    // Configure a new file-table entry and necessary state
+    int which = (conf.uart_base == UART0)? 0:
+                (conf.uart_base == UART1)? 1:
+                (conf.uart_base == UART2)? 2:
+                (conf.uart_base == UART3)? 3:
+                (conf.uart_base == UART4)? 4:5;
+
+    queue_init(&txq_headers[which], &txqs[which], 255);
+    queue_init(&rxq_headers[which], &rxqs[which], 255);
+
+    // make isr available

```

Comment:
whats going on with `which_uart` and `isr_active` here?

Reply:
e r g, I'll fix it